cpp/ql/src/Critical/OverflowStatic.ql

```diff
@@ -82,22 +82,31 @@ class CallWithBufferSize extends FunctionCall
   Expr buffer() {
     exists(int i |
       bufferAndSizeFunction(this.getTarget(), i, _) and
-      result = this.getArgument(i))
+      result = this.getArgument(i)
+    )
   }
-  Expr statedSize() {
+  Expr statedSizeExpr() {
     exists(int i |
       bufferAndSizeFunction(this.getTarget(), _, i) and
-      result = this.getArgument(i))
+      result = this.getArgument(i)
+    )
+  }
+  int statedSizeValue() {
+    exists(Expr statedSizeSrc |
+      DataFlow::localFlowStep*(DataFlow::exprNode(statedSizeSrc), DataFlow::exprNode(statedSizeExpr())) and
+      result = statedSizeSrc.getValue().toInt()
+    )
   }
 }
 
 predicate wrongBufferSize(Expr error, string msg) {
-  exists(CallWithBufferSize call, int bufsize, Variable buf |
+  exists(CallWithBufferSize call, int bufsize, Variable buf, int statedSize |
     staticBuffer(call.buffer(), buf, bufsize) and
-    call.statedSize().getValue().toInt() > bufsize and
-    error = call.statedSize() and
+    statedSize = call.statedSizeValue() and
+    statedSize > bufsize and
+    error = call.statedSizeExpr() and
```

Comment:
Now that you've introduced data flow, we need to decide what to do when multiple values may flow to `error`. With this PR we'll start seeing false positives in code like

```
char buf10[10], buf20[20];
if (b) { size = 10; } else { size = 20; }
if (b) {
  memcpy(buf10, src, size); // false positive
} else {
  memcpy(buf20, src, size);
}
```

You can be more conservative, and still address CPP-299, by replacing `call.statedSizeValue()` with `min(call.statedSizeValue())`. That won't address https://github.com/MosheWagner/VulnC/blob/f787c51fbe09d2fb88403d3b01bca7709fd7ed9e/src/bad_read.c#L158-L169, but I would prefer to err on the side of false negatives until we run a CPP-Differences job and see that such FPs are rare in practice.

Reply:
OK, makes sense to be conservative at this time.  Done.